torchao/csrc/cuda/fp6_llm/fp6_linear.cu

```diff
@@ -153,6 +156,39 @@ cudaError_t fpx_linear_kernel(cudaStream_t    stream,
 #include <ATen/cuda/CUDAContext.h>
 #include <torch/library.h>
 
+// https://github.com/NVIDIA/apex/blob/master/csrc/type_shim.h
+#if __CUDA_ARCH__ == 750
```

Comment:
We don't support <sm75 so equal is fine I think. Another look at this, maybe we can add this before BF16 case instead of having 2 `DISPATCH_HALF_AND_BF16`? something like

```C++
#if __CUDA_ARCH__ >= 800  // do we need to add ifdef ?
  case at::ScalarType::BFloat16: {                                             \
    using torch_t = at::BFloat16;                                              \
    using nv_t = __nv_bfloat16;                                                \
    __VA_ARGS__();                                                             \
    break;                                                                     \
  } 
#endif
```

Reply:
Ok on second look I think this check for `__CUDA_ARCH__` is actually just flat out wrong here. It's executed as host code, which means `__CUDA_ARCH__` will not be defined and so `#if __CUDA_ARCH__ == 750` will always be false, even when a sm_75 GPU is used. I verified this by checking for `#if __CUDA_ARCH__ >= 800` to define the macro, and that fails on a sm_80 GPU. A better solution would be to check for `__CUDA_ARCH__` further downstream in device code.

Reply:
In fact, I think any check for `__CUDA_ARCH__` in `fp6_linear.cu` will always fail because  `__CUDA_ARCH__` is undefined since all of the functions in `fp6_linear.cu` are host functions. So that means that the following guard (which was already there from previous iterations of the kernel) is also redundant since `__CUDA_ARCH__` is never defined here:

https://github.com/pytorch/ao/blob/aa86005df445ba1221523a3edc31f869a7da4f1c/torchao/csrc/cuda/fp6_llm/fp6_linear.cu#L21

E.g. if I change this guard to something that should always fail if `__CUDA_ARCH__` is defined:
```cpp
#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 99999
```
then it still evaluates to true on an A100 sm_80 GPU.

Reply:
I'm removing these guards for now, but it would be a good idea to also test this on sm75 to confirm that nothing unexpected happens. I might have time to test this on a Lightning Studio T4 in the next few days.

Reply:
I was just suggesting, didn't know if nested macro will work or not.

It seems like `cuda_bf16.h` is always available on host code, even for sm<75. So for `DISPATCH_HALF_AND_BF16`, we might not need to care about CUDA_ARCH at all?

A mini test by me on google colab

<img width="821" alt="image" src="https://github.com/user-attachments/assets/ff97ba08-002d-4b65-8479-bd7d2e768d31">

Reply:
Thanks a lot for testing this @gau-nernst. The actual usage of the __nv_bfloat16 type is still guarded by `__CUDA_ARCH__`, so this means I can probably remove the guards for the `cuda_bf16.h` include as well as the `DISPATCH_HALF_AND_BF16` macro. I'll do a sm75 test to confirm.

Reply:
Correction: It does seem to be necessary to guard against sm<=75 on device code, e.g. as done in the [Pytorch codebase](https://github.com/pytorch/pytorch/blob/ceab24def4605019666fc6d9aba7df6724732784/aten/src/ATen/native/cuda/KernelUtils.cuh#L4). But on host code it should be okay, like you tested.

Reply:
Correction 2: I tested on SM75 and the macro _does_ need to be guarded for SM75, otherwise it tries to compile device code with `nv_bfloat16` template type and error out since this is only available for SM>=80.